File: 1-refresh-token-strategy/backend/0-typescript/src/modules/auth/auth.controller.ts
/**
 * HTTP controller exposing the four auth endpoints used in the lesson flows:
 *  POST /auth/signup   — register a new user
 *  POST /auth/signin   — authenticate and receive AT+RT pair
 *  POST /auth/refresh  — rotate tokens (protected by RtGuard)
 *  POST /auth/logout   — revoke RT (protected by AtGuard)
 *
 * The controller is intentionally thin — all business logic lives in AuthService.
 */
import {
    Body,
    Controller,
    HttpCode,
    HttpStatus,
    Post,
    Req,
    UseGuards,
} from "@nestjs/common"
import {
    AtGuard,
    RtGuard,
} from "../../common"
import {
    AuthService,
} from "./auth.service"
import {
    SignInDto,
    SignUpDto,
} from "./dto"

/**
 * Provides the public auth API surface.
 * The "auth" prefix is resolved by the global prefix set in bootstrap.ts.
 */
@Controller("auth")
export class AuthController {
    constructor(private readonly authService: AuthService) {}

    /**
     * Registers a new account.
     * Returns HTTP 201 with `{ id, email }` — no tokens issued at signup.
     *
     * @param body - signup payload validated by SignUpDto
     */
    @Post("signup")
    @HttpCode(HttpStatus.CREATED) // explicit 201 — NestJS default POST is 201 but declared for clarity
    signUp(@Body() body: SignUpDto) {
        return this.authService.signUp(body)
    }

    /**
     * Authenticates credentials and issues an AT+RT pair.
     * Returns HTTP 200 with `{ access_token, refresh_token }`.
     *
     * @param body - signin payload validated by SignInDto
     */
    @Post("signin")
    @HttpCode(HttpStatus.OK) // override default 201 to match lesson contract (200)
    signIn(@Body() body: SignInDto) {
        return this.authService.signIn(body)
    }

    /**
     * Rotates the token pair given a valid refresh token in the Authorization header.
     * RtGuard runs RtStrategy which validates the JWT signature and attaches
     * `req.user = { sub, refreshToken }` before this handler executes.
     *
     * @param req - request with user populated by RtGuard/RtStrategy
     * @returns new `{ access_token, refresh_token }` pair
     */
    @Post("refresh")
    @HttpCode(HttpStatus.OK)
    @UseGuards(RtGuard) // validates RT signature; populates req.user.sub + req.user.refreshToken
    refresh(@Req() req: { user: { sub: number; refreshToken: string } }) {
        // Pass both the user id (for DB lookup) and raw token string (for bcrypt comparison).
        return this.authService.refreshTokens(req.user.sub,
            req.user.refreshToken)
    }

    /**
     * Revokes the active refresh token by setting its hash to null in the DB.
     * Requires a valid access token so the server can identify which user to log out.
     * AtGuard validates the AT and sets `req.user.userId`.
     *
     * @param req - request with user populated by AtGuard/AtStrategy
     * @returns `{ message: "Logged out" }`
     */
    @Post("logout")
    @HttpCode(HttpStatus.OK)
    @UseGuards(AtGuard) // validates AT; ensures only authenticated users can trigger revocation
    logout(@Req() req: { user: { userId: number } }) {
        return this.authService.logout(req.user.userId)
    }
}

File: 1-refresh-token-strategy/backend/1-java/src/main/java/com/example/refreshdemo/controller/AuthController.java
package com.example.refreshdemo.controller;

import com.example.refreshdemo.dto.SignInRequest;
import com.example.refreshdemo.dto.SignUpRequest;
import com.example.refreshdemo.dto.SignUpResponse;
import com.example.refreshdemo.dto.TokenResponse;
import com.example.refreshdemo.service.AuthService;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.validation.Valid;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.server.ResponseStatusException;

import java.util.HashMap;
import java.util.Map;

/**
 * REST controller exposing the four auth endpoints for the refresh-token strategy lesson.
 * The controller is intentionally thin — all business logic lives in {@link AuthService}.
 *
 * <ul>
 *   <li>POST /auth/signup   — register; returns {@link SignUpResponse}</li>
 *   <li>POST /auth/signin   — authenticate; returns {@link TokenResponse}</li>
 *   <li>POST /auth/refresh  — rotate tokens (RT in Authorization: Bearer header)</li>
 *   <li>POST /auth/logout   — revoke RT (AT required; principal set by JwtAuthenticationFilter)</li>
 * </ul>
 */
@RestController
@RequestMapping("/auth")
public class AuthController {

    /** Delegates all business logic (hashing, DB writes, token issuance). */
    private final AuthService authService;

    /**
     * @param authService injected auth service bean
     */
    public AuthController(AuthService authService) {
        this.authService = authService;
    }

    /**
     * Registers a new user with a BCrypt-hashed password.
     * Returns HTTP 201 with {@link SignUpResponse} containing id + email.
     * Returns HTTP 409 when the email is already registered.
     *
     * @param request validated signup payload (email + plaintext password)
     * @return new user's public profile (id and email)
     */
    @PostMapping("/signup")
    @ResponseStatus(HttpStatus.CREATED) // explicit 201 to match the lesson contract
    public SignUpResponse signUp(@Valid @RequestBody SignUpRequest request) {
        return authService.signUp(request);
    }

    /**
     * Authenticates credentials and issues a new AT + RT pair.
     * Returns HTTP 200 with {@link TokenResponse} containing both token strings.
     * Returns HTTP 401 when credentials are invalid.
     *
     * @param request validated signin payload (email + plaintext password)
     * @return token pair with access and refresh tokens
     */
    @PostMapping("/signin")
    public TokenResponse signIn(@Valid @RequestBody SignInRequest request) {
        return authService.signIn(request);
    }

    /**
     * Rotates the token pair given a valid refresh token in the Authorization header.
     * Two-step verification occurs in {@link AuthService#refreshTokens}: JWT signature/expiry,
     * then BCrypt hash comparison against the stored hash.
     * Returns HTTP 200 with a fresh {@link TokenResponse}.
     * Returns HTTP 401 when the token is expired, revoked, or the hash mismatches.
     *
     * @param request HTTP request (Authorization: Bearer header is read manually)
     * @return {@link ResponseEntity} containing a new token pair or a 401 error body
     */
    @PostMapping("/refresh")
    public ResponseEntity<?> refresh(HttpServletRequest request) {
        // Extract the raw RT string from the Authorization: Bearer header.
        String authHeader = request.getHeader("Authorization");
        if (authHeader == null || !authHeader.startsWith("Bearer ")) {
            return buildErrorResponse(HttpStatus.UNAUTHORIZED, "Refresh token revoked or rotated", "Unauthorized");
        }
        String refreshToken = authHeader.substring(7); // strip "Bearer " prefix (7 chars)
        try {
            TokenResponse tokens = authService.refreshTokens(refreshToken);
            return ResponseEntity.ok(tokens);
        } catch (ResponseStatusException e) {
            // Re-map the service exception to the lesson-spec JSON error body.
            return buildErrorResponse(HttpStatus.valueOf(e.getStatusCode().value()), e.getReason(), "Unauthorized");
        }
    }

    /**
     * Revokes the active refresh token by setting its BCrypt hash to null in the DB.
     * The access token is validated upstream by {@code JwtAuthenticationFilter};
     * the user id is read from the Spring Security {@code SecurityContext}.
     * Returns HTTP 200 with {@code { "message": "Logged out" }}.
     * Returns HTTP 401 when the principal is missing or anonymous.
     *
     * @return {@link ResponseEntity} with logout confirmation or 401 error
     */
    @PostMapping("/logout")
    public ResponseEntity<?> logout() {
        // The AT was already validated by JwtAuthenticationFilter; read the authenticated principal.
        Object principal = SecurityContextHolder.getContext().getAuthentication().getPrincipal();
        if (principal == null || "anonymousUser".equals(principal)) {
            // Should not happen if SecurityConfig is correct, but guard defensively.
            return buildErrorResponse(HttpStatus.UNAUTHORIZED, "Unauthorized", "Unauthorized");
        }
        // Principal is stored as String (user id) by JwtAuthenticationFilter.
        Long userId = Long.valueOf((String) principal);
        authService.logout(userId);

        Map<String, String> response = new HashMap<>();
        response.put("message", "Logged out");
        return ResponseEntity.ok(response);
    }

    /**
     * Builds a JSON error response body that matches the lesson HTTP contract:
     * {@code { "statusCode": <N>, "message": "...", "error": "..." }}.
     *
     * @param status  HTTP status code to set on the response
     * @param message human-readable error description
     * @param error   error category string (e.g. "Unauthorized")
     * @return {@link ResponseEntity} with the error body and the given status
     */
    private ResponseEntity<?> buildErrorResponse(HttpStatus status, String message, String error) {
        Map<String, Object> body = new HashMap<>();
        body.put("statusCode", status.value());
        body.put("message", message);
        body.put("error", error);
        return ResponseEntity.status(status).body(body);
    }
}
